src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs

```diff
@@ -33,9 +35,36 @@ public abstract class FormulaType
         public static FormulaType OptionSetValue { get; } = new OptionSetValueType();
 
         // chained by derived type 
-        internal FormulaType(DType type)
+        internal FormulaType(DType type, Dictionary<string, string> displayNameSet = null)
         {
-            _type = type;
+            if (displayNameSet == null)
+                _type = type;
+            else 
+            {
+                ValidateDisplayNames(displayNameSet);
+                _type = DType.AttachDisplayNameProvider(type, new BidirectionalDictionary<string, string>(displayNameSet));
+            }
+
+        }
+
+        private void ValidateDisplayNames(Dictionary<string, string> displayNameSet)
+        {
+            HashSet<string> displayNames = new();
+            foreach (var kvp in displayNameSet)
+            {
+                // Display names can't collide with other display names
+                if (displayNames.Contains(kvp.Value))
+                {
+                    throw new DisplayNameCollisionException(kvp.Value);
```

Comment:
Should we be throwing an exception on the first display name conflict? What if there are multiple conflicts?

Reply:
Changed API to make this clearer